multinet/uploaders/d3_json.py

```diff
@@ -0,0 +1,99 @@
+"""Multinet uploader for nested JSON files."""
+import json
+from io import StringIO
+import pandas as pd
+from collections import OrderedDict
+
+from .. import db, util
+from ..errors import ValidationFailed
+from ..util import decode_data
+
+from flask import Blueprint, request
+from flask import current_app as app
+
+# Import types
+from typing import Any
+
+bp = Blueprint("d3_json", __name__)
+bp.before_request(util.require_db)
+
+
+def validate_d3_json(data):
+    """Perform any necessary d3 json validation, and return appropriate errors."""
+    data_errors = []
+
+    # Check the structure of the uploaded file is what we expect
+    if "nodes" not in data.keys() or "links" not in data.keys():
+        data_errors.append({"error": "structure"})
+
+    # Check that links are in source -> target form
+    if not all(
+        "source" in row.keys() and "target" in row.keys() for row in data["links"]
+    ):
+        data_errors.append({"error": "invalid_link_keys"})
+
+    # Check that the keys for each dictionary match
+    if not all(data["links"][0].keys() == row.keys() for row in data["links"]):
+        data_errors.append({"error": "inconsistent_link_keys"})
+
+    # Check for duplicated nodes
+    ids = [row["id"] for row in data["nodes"]]
+    if len(ids) != len(set(ids)):
+        data_errors.append({"error": "node_duplicates"})
+
+    # Duplicated edges are okay, they might encode different relationships, but
+    # that should be specified in the attributes. Thus duplicated dicts are not okay
+    if len(data["links"]) != len(set([tuple(x.items()) for x in data["links"]])):
+        data_errors.append({"error": "link_duplicates"})
```

Comment:
I'm not sure I understand this check--I think it's checking for identical attributes among different links, right? But why wouldn't that be allowed?

Reply:
This is checking for duplicate links by converting the dict into a tuple and comparing them all. Duplicate links shouldn't be allowed, they should be collapsed down into one object

Reply:
Right, but why can't we have multiple links with the same data on them?

In the simplest case, there wouldn't be any data on the link at all, and the multiplicity itself is the desired state.

Reply:
I was thinking that in the case that you have the following link:

```
{source: a, target: b, likes: 5}
```

There should not be duplicates allowed. I do see your point about potentially allowing the duplicates, too. Should I remove this code and we can leave it to the user to properly format their data?

Reply:
I'm almost certain that if you load a d3 json file with a repeated link record like that, it will parse just fine in, e.g., d3's own graph vis libraries. There may certainly be applications where you would want to restrict this sort of thing, but there are definitely applications where this would be expected. For that reason I don't believe we should disallow it at this most general level.

Reply:
Removed this code in af0ee12